fastparquet/speedups.pyx

```diff
@@ -4,164 +4,149 @@ Native accelerators for Parquet encoding and decoding.
 
 from __future__ import absolute_import
 
-cdef extern from "string.h":
-    void *memcpy(void *dest, const void *src, size_t n)
-
-from cpython cimport (PyUnicode_AsUTF8String, PyUnicode_DecodeUTF8,
-                      PyBytes_CheckExact, PyBytes_FromStringAndSize,
-                      PyBytes_GET_SIZE, PyBytes_AS_STRING)
-
+import array
 import numpy as np
 cimport numpy as np
-
-
+import cython
+cimport cython
+import sys
+from cpython cimport (array, PyBytes_GET_SIZE, PyBytes_AS_STRING, PyBytes_Check,
+                      PyBytes_FromStringAndSize, PyUnicode_AsUTF8String)
+from cpython.buffer cimport (PyBUF_ANY_CONTIGUOUS, PyObject_GetBuffer,
+                             PyBuffer_Release)
+from libc.string cimport memcpy
+from libc.stdlib cimport malloc, free
+
+PY2 = sys.version_info[0] == 2
 _obj_dtype = np.dtype('object')
 
+cdef extern from "Python.h":
+    bytearray PyByteArray_FromStringAndSize(char *v, Py_ssize_t l)
+    char* PyByteArray_AS_STRING(object string)
+    object PyUnicode_FromStringAndSize(const char *u, Py_ssize_t size)
+    int PyUnicode_Check(object text)
+    int PyBytes_AsStringAndSize(bytes obj, char **buffer, Py_ssize_t *length)
+    char* PyUnicode_AsUTF8AndSize(unicode unicode, Py_ssize_t *size)
+    unicode PyUnicode_DecodeUTF8(char *s, Py_ssize_t size, char *errors)
 
-def _to_array(obj):
-    """
-    Convert *obj* (a ndarray-compatible object, e.g. pandas Series)
-    to a true ndarray.
-    """
-    if not isinstance(obj, np.ndarray):
-        try:
-            obj = obj.__array__()
-        except AttributeError:
-            raise TypeError("expected an ndarray-compatible object, got %r"
-                            % (type(obj,)))
-        assert isinstance(obj, np.ndarray)
-    return obj
-
-
-def _check_1d_object_array(np.ndarray arr):
-    if arr.ndim != 1:
-        raise TypeError("expected a 1d array")
-    if arr.dtype != _obj_dtype:
-        raise TypeError("expected an object array")
-
-
-def array_encode_utf8(inp):
-    """
-    utf-8 encode all elements of a 1d ndarray of "object" dtype.
-    A new ndarray of bytes objects is returned.
-    """
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+def encode(buf, bint utf8=1):
     cdef:
-        Py_ssize_t i, n
-        np.ndarray[object] arr
-        np.ndarray[object] result
+        Py_ssize_t i, l, n_items, data_length, total_length
+        object[:] input_values
+        char** encoded_values
+        int[:] encoded_lengths
+        char* encv
+        char* b
+        bytearray out
+        char* data
+        object u
+
+    # normalise input
+    input_values = np.asanyarray(buf, dtype=object).reshape(-1, order='A')
+
+    # determine number of items
+    n_items = input_values.shape[0]
+
+    # setup intermediates
+    encoded_values = <char **>malloc(n_items * sizeof(char *))
```

Comment:
You should add a `try...finally` to ensure `free()` is called in case of exception (or find another strategy :-)).